e2e/cypress/tests/integration/performance/channel_switch_spec.js

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+// ***************************************************************
+// - [#] indicates a test step (e.g. # Go to a page)
+// - [*] indicates an assertion (e.g. * Check the title)
+// - Use element ID when selecting an element. Create one if none.
+// ***************************************************************
+
+// Stage: @prod
+// Group: @channel
+
+describe('Channel switch performance test', () => {
+    let testUser;
+
+    before(() => {
+        cy.apiInitSetup().then(({team, user}) => {
+            testUser = user;
+
+            // # Login as test user and go to town square
+            cy.apiLogin(testUser);
+            cy.visit(`/${team.name}/channels/town-square`);
+        });
+    });
+
+    it('measures switching between two channels from LHS', () => {
+        // # Invoke window object
+
+        cy.window().
+            its('performance').then((performance) => {
+                // # Switch channel to Off-topic
+
+                cy.get('#sidebarItem_off-topic').click({force: true}).
+
+                // # Mark end of the load
+
+                    then(() => performance.mark('channelLoad')).
+                    then(() => {
+                        performance.measure('channelLoad');
+                        const measure = performance.getEntriesByName('channelLoad')[0];
+                        const duration = measure.duration;
+
+                        // * Verify the duration is less than the specified amount
+                        assert.isAtMost(duration, 5000);
```

Comment:
The problem of using one single measure to decide whether it is good or not is that it becomes too brittle. Imagine there are some uncontrollable circumstances that make this test to go faster than usual or slower than usual. That would make you pass or reject this test when the problem is not the performance, but some random variance.

If, to avoid that, we make the maximum duration too big, what we lose is granularity. Even if there is a loss in performance, if it does not get over the threshold, you won't notice.

Therefore, we would need to run this test several time, and use statistics to make sure the difference is relevant.

Reply:
@nevyangelova cc: @larkox 

What about something along the lines of a reusable way that can be provided to everyone as a helper to wrap the functionality that Daniel suggests eg (not a complete code example below, but something along those lines)?

```js
const measure = async (name, upperLimit, cb, runs=10) => {
    return cy.window().its('performance').then(async (performance) => {
        performance.mark(name);

        for(let i = 0; i < runs; i++) {
            await cb();
        }
    
        performance.measure(name);
        const measure = performance.getEntriesByName(name)[0];
        const durationMs = Math.round(measure.duration/runs);

        cy.log('[PERFORMANCE] ${name}: ${durationMs}ms', durationMs <= upperLimit ? ' within upper limit' : ' outside upper limit', ` of ${upperLimit}ms`);
    
        assert.isAtMost(durationMs, upperLimit);    
    });
};


describe('Channel switch performance test', () => {
    let testUser;

    before(() => {
        cy.apiInitSetup().then(({team, user}) => {
            testUser = user;

            // # Login as test user and go to town square
            cy.apiLogin(testUser);
            cy.visit(`/${team.name}/channels/town-square`);
        });
    });

    it('measures switching between two channels from LHS', async () => {
        // # Invoke window object
        await measure('channelLoad', 1000, () => {
            cy.get('#sidebarItem_off-topic').click({force: true});
        });
    });
});

```

Reply:
Nice one @m1lt0n currently testing with this proposed function, will update the results now but its a very good idea!